x/subscription/keeper/cu_tracker.go

```diff
@@ -5,34 +5,188 @@ import (
 	"strconv"
 
 	sdk "github.com/cosmos/cosmos-sdk/types"
+	legacyerrors "github.com/cosmos/cosmos-sdk/types/errors"
 	"github.com/lavanet/lava/utils"
+	dualstakingtypes "github.com/lavanet/lava/x/dualstaking/types"
 	"github.com/lavanet/lava/x/subscription/types"
 )
 
-func (k Keeper) GetTrackedCu(ctx sdk.Context, sub string, provider string, chainID string, block uint64) (cu uint64, found bool, key string) {
+// GetTrackedCu gets the tracked CU counter (with/without QoS influence) and the trackedCu entry's block
+func (k Keeper) GetTrackedCu(ctx sdk.Context, sub string, provider string, chainID string) (cu uint64, entryBlock uint64, found bool, key string) {
 	cuTrackerKey := types.CuTrackerKey(sub, provider, chainID)
 	var trackedCu types.TrackedCu
-	found = k.cuTrackerFS.FindEntry(ctx, cuTrackerKey, block, &trackedCu)
-	return trackedCu.GetCu(), found, cuTrackerKey
+	entryBlock, found = k.cuTrackerFS.FindEntry2(ctx, cuTrackerKey, uint64(ctx.BlockHeight()), &trackedCu)
+	if !found {
+		// entry not found -> this is the first, so not an error. return the current block and CUs=0
+		return 0, uint64(ctx.BlockHeight()), found, cuTrackerKey
+	}
+	return trackedCu.Cu, entryBlock, found, cuTrackerKey
 }
 
-func (k Keeper) AddTrackedCu(ctx sdk.Context, sub string, provider string, chainID string, cu uint64, block uint64) error {
-	if sub == "" || provider == "" || block > uint64(ctx.BlockHeight()) {
-		return utils.LavaFormatError("cannot add tracked CU",
-			fmt.Errorf("sub/provider cannot be empty. block cannot be larger than current block"),
+// AddTrackedCu adds CU to the CU counters in relevant trackedCu entry
+// Note that the trackedCu entry always has one version (updating it using append in the same block acts as ModifyEntry)
+func (k Keeper) AddTrackedCu(ctx sdk.Context, sub string, provider string, chainID string, cuToAdd uint64) error {
+	if sub == "" || provider == "" {
```

Comment:
how would this be the case?

Reply:
done. see https://github.com/lavanet/lava/commit/93dc7f6261cac796b09559e5fa4c33be54361ac4